qualtran/simulation/tensor/tensor_from_classical.py

```diff
@@ -0,0 +1,136 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import itertools
+from typing import Iterable, TYPE_CHECKING
+
+import numpy as np
+from numpy.typing import NDArray
+
+if TYPE_CHECKING:
+    import quimb.tensor as qtn
+
+    from qualtran import Bloq, ConnectionT, Register
+    from qualtran.simulation.classical_sim import ClassicalValT
+
+
+def _bits_to_classical_reg_data(reg: 'Register', bits) -> 'ClassicalValT':
+    if reg.shape == ():
+        return reg.dtype.from_bits(bits)
+    return reg.dtype.from_bits_array(np.reshape(bits, reg.shape + (reg.dtype.num_qubits,)))
+
+
+def _bloq_to_dense_via_classical_action(bloq: 'Bloq') -> NDArray:
+    """Internal method to compute the tensor of a bloq using its classical action.
+
+    Args:
+        bloq: the Bloq
+
+    Returns:
+        an NDArray of shape (2, 2, ...) indexed by the output bits followed by input bits.
+    """
+    left_qubit_counts = tuple(reg.total_bits() for reg in bloq.signature.lefts())
+    left_qubit_splits = np.cumsum(left_qubit_counts)
+
+    n_qubits_left = sum(left_qubit_counts)
+    n_qubits_right = sum(reg.total_bits() for reg in bloq.signature.rights())
+
+    matrix = np.zeros((2,) * (n_qubits_right + n_qubits_left))
```

Comment:
consider adding a check with a helpful error message if n_qubits_right + n_qubits_left is an unreasonably large number

Reply:
raises a value error when total > 40